### PR TITLE
Warn if base of exponentiation operation is a literal.

### PR DESCRIPTION
**Pull request by chriseth**
Fixes https://github.com/ethereum/solidity/issues/1745

### CODE REVIEW

**Comment by axic** [comment]
I think it would be useful to add end-to-end tests with some truncated values (to ensure the type selection is correct in every case).

File: libsolidity/analysis/TypeChecker.cpp [hunk 1]
```diff
@@ -1092,6 +1092,26 @@ void TypeChecker::endVisit(BinaryOperation const& _operation)
 		Token::isCompareOp(_operation.getOperator()) ?
 		make_shared<BoolType>() :
 		commonType;
+	if (_operation.getOperator() == Token::Exp)
+	{
+		if (
+			leftType->category() == Type::Category::RationalNumber &&
+			rightType->category() != Type::Category::RationalNumber
```
Comment (pirapira) [inline libsolidity/analysis/TypeChecker.cpp:1099]: Why cannot the right type be a rational?

Comment (chriseth) [reply libsolidity/analysis/TypeChecker.cpp:1099]: In this case, the result is also a rational and there is on precision loss.

Comment (pirapira) [reply libsolidity/analysis/TypeChecker.cpp:1099]: The result type is anyway checked below, but OK, it doesn't hurt.

**Review by pirapira — APPROVED** [review]
Looks good to me.

**Comment by axic** [comment]
In the end tests weren't added for truncation.

**Comment by chriseth** [comment]
@axic sorry, this was kind of overlooked. What do you mean by "truncated values"? Can you give an example test?

**Pull request merged**